jax/_src/test_warning_util.py

```diff
@@ -27,106 +27,130 @@
+      log.append(warnings.WarningMessage(message, category, filename, lineno, file, line))
+      return True
 
-  log = []
+    _context.handlers.append(handler)
+    try:
+      yield log
+    finally:
+      _context.handlers.pop()
+
+
+  @contextlib.contextmanager
+  def ignore_warning(*, message: str | None = None, category: type = Warning):
+    "Context manager that ignores any matching warnings."
+    if warnings.showwarning is not _showwarning:
+      with warnings.catch_warnings():
+        warnings.filterwarnings(
+          "ignore", message="" if message is None else message, category=category)
+        yield
+      return
 
-  def handler(message, category, filename, lineno, file=None, line=None):
-    log.append(warnings.WarningMessage(message, category, filename, lineno, file, line))
-    return True
+    if message:
+      message_re = re.compile(message)
+    else:
+      message_re = None
 
-  _context.handlers.append(handler)
-  try:
-    yield log
-  finally:
-    _context.handlers.pop()
+    category_cls = category
 
+    def handler(message, category, filename, lineno, file=None, line=None):
+      text = str(message) if isinstance(message, Warning) else message
+      if (message_re is None or message_re.match(text)) and issubclass(
+          category, category_cls
+      ):
+        return True
+      return False
 
-@contextlib.contextmanager
-def ignore_warning(*, message: str | None = None, category: type = Warning):
-  "Context manager that ignores any matching warnings."
-  if warnings.showwarning is not _showwarning:
-    with warnings.catch_warnings():
-      warnings.filterwarnings(
-        "ignore", message="" if message is None else message, category=category)
+    _context.handlers.append(handler)
+    try:
       yield
-    return
+    finally:
+      _context.handlers.pop()
 
-  if message:
-    message_re = re.compile(message)
-  else:
-    message_re = None
 
-  category_cls = category
+  def install_threadsafe_warning_handlers():
+    # Hook the showwarning method. The warnings module explicitly notes that
+    # this is a function that users may replace.
+    warnings.showwarning = _showwarning
 
-  def handler(message, category, filename, lineno, file=None, line=None):
-    text = str(message) if isinstance(message, Warning) else message
-    if (message_re is None or message_re.match(text)) and issubclass(
-        category, category_cls
-    ):
-      return True
-    return False
+    # Set the warnings module to always display warnings. We hook into it by
+    # overriding the "showwarning" method, so it's important that all warnings
+    # are "shown" by the usual mechanism.
+    warnings.simplefilter("always")
 
-  _context.handlers.append(handler)
-  try:
-    yield
-  finally:
-    _context.handlers.pop()
 
+else:
+  @contextlib.contextmanager
+  def raise_on_warnings():
+    with warnings.catch_warnings(action="error"):
```

Comment:
Could we just `return warnings.catch_warnings(action="error")` here?

Reply:
No, it seems not:
```
Traceback (most recent call last):
  File "/Users/phawkins/p/jax/tests/warnings_util_test.py", line 26, in <module>
    class WarningsUtilTest(jtu.JaxTestCase):
    ...<55 lines>...
              warnings.warn("hello", category=UserWarning)
  File "/Users/phawkins/p/jax/tests/warnings_util_test.py", line 28, in WarningsUtilTest
    @test_warning_util.raise_on_warnings()
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
TypeError: 'catch_warnings' object is not callable
```